fix: generate random OAuth state each login

ProgrammerSONGBIT opened the pull request:
## Changes
- Added `generateRandomState()` using `crypto.getRandomValues()` for cryptographically secure random OAuth state
- Modified `handleLogin()` to append unique `&state=${randomState}` to `GITHUB_OAUTH_URL` each login
- Each "Student/Mentor Sign-up" click now generates **different 43-char state parameter**

## Why this matters
Prevents OAuth CSRF attacks by ensuring `state` parameter is unique per login attempt [web:162]

## Testing
- Click "Student Sign-up" → URL gets `&state=abcXYZ123random...`
- Refresh → Click again → **NEW** `&state=def456different...`

ProgrammerSONGBIT committed:
fix: generate random OAuth state each login

Fixes #216

## Changes
- Added `generateRandomState()` using `crypto.getRandomValues()` for cryptographically secure random OAuth state
- Modified `handleLogin()` to append unique `&state=${randomState}` to `GITHUB_OAUTH_URL` each login
- Each "Student/Mentor Sign-up" click now generates **different 43-char state parameter**

## Why this matters
Prevents OAuth CSRF attacks by ensuring `state` parameter is unique per login attempt [web:162]

## Testing
- Click "Student Sign-up" → URL gets `&state=abcXYZ123random...`
- Refresh → Click again → **NEW** `&state=def456different...`

dipamsen reviewed (changes requested):
This PR does not completely implement #246. It generates a random state for logging in, but does not verify whether the state is correct after the redirect. See the description in the issue for more details.


> Solution:
> 1. Generate a random string when clicking on the login button
> 2. Store the string in `localstorage` and redirect to the oauth URL with this state
> 3. When the oauth login is complete and the user is redirected to `/oauth` on the frontend, match these strings and only proceed to log in if they match
> 4. It would also be good to make the state a limited-time string so that a string generated previously cannot be used

ProgrammerSONGBIT committed:
Enhance handleLogin with localStorage for OAuth

Added localStorage handling for OAuth state and timestamp.

ProgrammerSONGBIT committed:
Implement OAuth state validation and cleanup

Added OAuth state validation to prevent CSRF attacks and handle state expiration.

dipamsen commented on src/pages/OAuth.tsx:
```diff
@@ -8,6 +8,61 @@ function OAuth() {
+
+    // Check 3: States match (prevents CSRF attacks)
+    if (receivedState !== storedState) {
+      console.error("CSRF ATTACK DETECTED: OAuth state mismatch!", {
```
No need to provide such descriptive error messages to the user (specifically in `setErr`) -- the error shown to the user can simply be 'Something went wrong. Please try logging in again.'

dipamsen commented on src/pages/OAuth.tsx:
```diff
@@ -33,22 +88,40 @@ function OAuth() {
+      return;
     }
+
+    // Validate OAuth state (NEW - prevents CSRF attacks)
```
please remove the comments

ProgrammerSONGBIT committed:
Comments are removed and added simple error message 

1. Added simple error message
2. Comments are removed

ProgrammerSONGBIT commented:
Removed comments as per your feedback and added simple error messages.

ProgrammerSONGBIT committed:
Improve error message for OAuth state mismatch

harshkhandeparkar commented on src/components/Home/HeroSection.tsx:
```diff
@@ -11,14 +11,30 @@ import { UserType } from "../../util/types";
-    auth.setUserType(userType as UserType);
-    window.location.href = GITHUB_OAUTH_URL;
-  };
+const handleLogin = (userType: string) => {
```
Instead of a separate function, can this be part of the auth context?

ProgrammerSONGBIT replied:
> Instead of a separate function, can this be part of the auth context?

Changed the auth tsx file

ProgrammerSONGBIT committed:
Refactor HeroSection to use auth context for login

Removed GitHub OAuth logic and replaced with auth context method.

ProgrammerSONGBIT committed:
Add handleOAuthLogin method to AuthContext

Added handleOAuthLogin method to AuthContext.

ProgrammerSONGBIT committed:
Implement OAuth login handler with CSRF protection

Added OAuth login handler and random state generation for CSRF protection in the auth file .

ProgrammerSONGBIT committed:
revert OAuth.tsx

ProgrammerSONGBIT committed:
revert HeroSection.tsx

ProgrammerSONGBIT committed:
revertprev OAuth.tsx

ProgrammerSONGBIT committed:
Delete src/components/Home/HeroSection.tsx

ProgrammerSONGBIT committed:
Delete src/pages/OAuth.tsx

ProgrammerSONGBIT committed:
Delete src/util/auth.tsx
